### DISCUSSION TITLE
LoRa - Switch from LoRaMac-node to LoRa basic modem as library planned

### DISCUSSION THREAD

**Discussion by marcoster** [zephyrproject-rtos/zephyr#72442] [General]
Hello there,

is there any plan to switch to the newer library?
It would bring support for the 2.4GHz Chips (SX128x) and is also the one in active development from Semtech.
LoRaMac-node is already 'only' in maintenance mode - so nothing new will come here...

I hope this is the right place to ask.

Thanks
Marco

**Comment by ndrs-pst** [comment]
I just realized this too. Maybe Zephyr `v4.0.0` would be great.
Ping @martinjaeger 😄

By the way, it looks like Semtech provides a decent porting guide at the following link. Maybe @mluis1, the author of LBM, can participate in this integration with Semtech's endorsement:
https://www.semtech.com/products/wireless-rf/lora-connect/sx1262#documentation

---

@marcoster I think this is what you are looking for https://github.com/IRNAS/SWL2001-Zephyr

**Comment by martinjaeger** [comment]
According to the Discord channel, @Salamandar from Semtech is already working on a port. Maybe he can give an update on the progress so far?

**Reply by Salamandar** [reply]
Hi @martinjaeger, sorry for the late reply.

Yes, indeed, I did port LoRa Basics Modem to Zephyr. Everything is working perfectly fine, and we will make a public release, I hope in the next few days.

I'm kinda sure my current code is not in a mergeable state though, but i'll be happy to work with you on that.
The reason is that my goal was to make the LoRaWAN stack LBM work. This stack has an "internal" API called the RAL (Radio Abstraction Layer, see https://github.com/Lora-net/SWL2001/blob/d05dad5b4df9c0e5899008e43bfce48f5498abb5/lbm_lib/smtc_modem_core/smtc_ral/src/ral_drv.h#L155), so I did not need to write a "public API" for the radio drivers. I think that the Zephyr community wants a public API, and it would need to be written.

Also the current Semtech transceivers have a very extensive API (see the RAL) and having multi-api devices would help (see https://github.com/zephyrproject-rtos/zephyr/issues/48934) :)

Anyways, I'll ping here once the code is made public and you can give your feedback ;)

**Reply by Salamandar** [reply]
I made a quick'n'dirty merge of my repository with Zephyr, and did a PR, clearly not merge-ready, but I think you guys will be happy to see it. It's not yet buildable as I'm waiting for a branch to be pushed in the official "SWL2001" repository with some patches. I'll also add some more explanations next week about what I think needs to be improved to be mergeable.
https://github.com/zephyrproject-rtos/zephyr/pull/77495

**Reply by rdagher** [reply]
This is very good news @Salamandar  !

For legacy hardware (gloups), what about the experimental  SX1272/SX1276 support. Any chance to see it work in zephyr too?

**Reply by Salamandar** [reply]
Yes, sx128x was not in the scope of my mission, but i'm open to work with you on that, now that code is public !

**Reply by Salamandar** [reply]
@martinjaeger I cleaned up a bit the branch (mostly : run CI, apply Zephyr coding style, etc), compliance now passes. 
Basically there are 3 questions at this point:
* I'm not sure about which samples to port/write
* I don't know which API to implement for the drivers (e.g raw drivers without LoRaWAN stack on top). Should I implement the API that the loramac-node implementation currently provides ?
* I'm not sure what to do with the storage API and Device Management callbacks (temperature/battery level/etc), the current code is basically crap.
I still have some LBM lorawan samples to fully port (they were written with Semtech coding style / porting APIs, i should cleanup those relevant to be ported).

Anyways, now that the PR is here, we can fill in the gaps :)

**Reply by g-levrel** [reply]
Hi thank you for your work, 

Having raw driver access would be useful as it'd allow simple LoRa communication without having to run the LoRaWan stack.

Support for the SX1280 would also be be very useful if it's possible.

**Comment by sluissantos** [comment]
That's great news! Just out of curiosity... how many days left? Haha
